source/util/AMTrapezoidVelocityProfile.cpp

```diff
@@ -1,49 +1,62 @@
 #include "AMTrapezoidVelocityProfile.h"
-
+#include <math.h>
 AMTrapezoidVelocityProfile::AMTrapezoidVelocityProfile()
 {
-
-    baseVelocity_ = 0;
-    initialAcceleration_ = 0;
-    targetVelocity_ = 0;
-    returnAcceleration_ = 0;
+	startPosition_ = 0;
+	endPosition_ = 0;
+	acceleration_ = 0;
+	time_ = 0;
+	isValid_ = false;
 }
 
-AMTrapezoidVelocityProfile::AMTrapezoidVelocityProfile(double baseVelocity,
-                                                       double initialAcceleration,
-                                                       double targetVelocity,
-                                                       double returnAcceleration)
+AMTrapezoidVelocityProfile::AMTrapezoidVelocityProfile(double startPosition,
+                                                       double endPosition,
+                                                       double acceleration,
+                                                       double time)
 {
+	startPosition_ = startPosition;
+	endPosition_ = endPosition;
+	acceleration_ = acceleration;
+	time_ = time;
+	isValid_ = true;
+}
 
-    baseVelocity_ = baseVelocity;
-    initialAcceleration_ = initialAcceleration;
-    targetVelocity_ = targetVelocity;
-    returnAcceleration_ = returnAcceleration;
+double AMTrapezoidVelocityProfile::startPosition() const
+{
+	return startPosition_;
 }
 
-double AMTrapezoidVelocityProfile::baseVelocity() const
+double AMTrapezoidVelocityProfile::endPosition() const
 {
-    return baseVelocity_;
+	return endPosition_;
 }
 
-double AMTrapezoidVelocityProfile::initialAcceleration() const
+double AMTrapezoidVelocityProfile::acceleration() const
 {
-    return initialAcceleration_;
+	return acceleration_;
 }
 
-double AMTrapezoidVelocityProfile::targetVelocity() const
+double AMTrapezoidVelocityProfile::time() const
 {
-    return targetVelocity_;
+	return time_;
 }
 
-double AMTrapezoidVelocityProfile::returnAcceleration() const
+double AMTrapezoidVelocityProfile::targetVelocity() const
 {
-    return returnAcceleration_;
+	double halfAccelByTime = (acceleration_*time_) / 2;
+	double toRoot = pow(halfAccelByTime, 2) - acceleration_ * qAbs(endPosition_ - startPosition_);
+
+	if(toRoot < 0) {
+		return halfAccelByTime + sqrt(qAbs(toRoot));
```

Comment:
This was arrived on with help from David. Essentially this bottom case is when the acceleration is not sufficient to reach the target velocity within the delta T. In that case this simply needs to ensure its a value above that peak, which this does.